### PR TITLE
feature: provide error listeners

### PR DESCRIPTION
**Pull request by nodify-at**
feature: provide error listeners for catalog processing engines to notify the callers if an entity can not be processed.

Signed-off-by: Hasan Oezdemir <21654050+nodify-at@users.noreply.github.com>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by nodify-at** [comment]
This feature would be helpful to track the entity processing errors. We would like to get notified if an error occurs to collect and extract the errors to our external systems. We are also planing to create automated issues if possible whenever an entity is not processable.

**Comment by manusant** [comment]
This is something that is really missing. Currently the only way to track issues is by checking the logs

**Review by jhaals — COMMENTED** [review]
Thanks for the suggestion!

It's quite clear that we need to do a better job at processing surfacing errors. In isolation this PR does take a step in the right direction. However this only makes it better for our users that goes through the effort of discovering this API and creates some custom reporting or instrumentation. I would like us to get to a point where this new functionality would be used in the frontend or elsewhere in the system as it's currently just a dangling tool that we expose and leave it up to the integrator to hook into it.
Taking this into the bigger picture there might also be some overlap with this and the event RFC #8219. At the very least I think these new types and endpoints should be clearly marked `@alpha` and say experimental until we know a bit more how this should work overall.

Would be great to get some additional input from @freben and @backstage/maintainers on this.

**Comment by nodify-at** [comment]
@jhaals Thanks for pointing to RFC. This sounds cool to track the events inside the catalog engine. Is there a plan for this feature? 

@freben thanks for sharing your feedback. One benefit to have a listener or an event listener directly the engine itself is to have a one entry point to track the errors otherwise we need to implement for each processor this handleError, but while this RFC is still in progress/review, we can use it. In the commit which you sent, I see this one is removed and never called by catalog engine, can I prepare a PR to rollback this commit and to call handleError inside the processing engine ?

**Comment by Rugvip** [comment]
@nodify-at the `handleError` doesn't quite capture the same scope of errors that I think you'd be looking for. Not really something we'd want to add back at this point either tbh.

As Jhaals hinted at this change is quite controversial because we don't know whether this is the way that we want to end up making errors available. This method in particular ends up leaving a lot to be implemented by consumers of the plugin. It also doesn't provide much in the way of delivery guarantees or persistence, both of which it's not very clear whether we need or not.

Something that would help move this forward is if you could elaborate in a bit more detail how you intend to use this. How do you plan to gather the errors, store them, surface them to users, etc

Just to move this forward for now, one tweak we know we'd like to do regardless is to make this available as an addition to the `CatalogBuilder` instead, and having that be the only public API addition. Do you think that would need a quick enough refactor, so that we can have a look at what that looks like?

**Comment by nodify-at** [comment]
@Rugvip Thanks for your feedback. I can try a quick refactoring so we can check together how does that looks like. I will push till Monday.

**Comment by freben** [comment]
@nodify-at I don't think the discussed changes were made yet, right? Do you have plans to keep working on this?

**Comment by nodify-at** [comment]
@Rugvip Sorry, I was little bit busy in last days, I'd like to refactor this to provide it via CatalogBuilder but can you give me a little bit more detail about it? You want to see this something like: Setting a listener via CatalogBuilder (e.g. builder.setErrorListener instead of passing it to processing engine directly?

**Comment by freben** [comment]
@nodify-at What we'd suggest is adding it to the `CatalogBuilder` as a new method, yeah. Maybe something like

```
subscribe(listeners: {
  processingError: (options: { entity: Entity, ... }) => void | Promise<void>,
})
```

so we can add more subscriber types in the future if we want to.

**Comment by nodify-at** [comment]
@freben Thanks for your feedback :) I have refactored the code, can you give me a feedback? I hope I understood the refactoring requirements correctly :)

**Comment by nodify-at** [comment]
@freben I improved the code based on your feedback, thanks! I've removed the typing we may add it later as you mentioned. Can you please check the changes? :)

File: plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts
Comment (nodify-at) [reply plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts:48]: Yes, one subscription should be fine for now :)

File: plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts
Comment (nodify-at) [reply plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts:45]: This is called at line 449 in CatalogBuilder, is it ok?

File: plugins/catalog-backend/api-report.md
Comment (nodify-at) [reply plugins/catalog-backend/api-report.md:424]: I didn't touch to EntityProcessingResult but I removed the @public and added @internal (also removed from index.ts)

**Comment by nodify-at** [comment]
@freben ready for next round :)

**Comment by nodify-at** [comment]
@freben would you have some time to review my latest changes?

**Review by freben — APPROVED** [review]
Nice!

**Pull request merged**
